In [105]:
from __future__ import division, print_function, absolute_import
from past.builtins import basestring

import os
import gzip

import pandas as pd

from twip.constant import DATA_PATH

from gensim.models import TfidfModel, LsiModel
from gensim.corpora import Dictionary

In [106]:
import matplotlib
from IPython.display import display, HTML 
%matplotlib inline
np = pd.np
display(HTML("<style>.container { width:100% !important; }</style>"))
pd.set_option('display.max_rows', 6)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 500)
pd.set_option('precision', 2)
%precision 4
%pprint

Pretty printing has been turned ON


In [150]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [151]:
lsi = LsiModel.load(os.path.join(DATA_PATH, 'lsi100'))
lsi2 = LsiModel.load(os.path.join(DATA_PATH, 'lsi2'))

In [152]:
with gzip.open(os.path.join(DATA_PATH, 'tweet_topic_vectors.csv.gz'), 'rb') as f:
    topics = pd.DataFrame.from_csv(f, encoding='utf8')
topics = topics.fillna(0)

In [153]:
topics

0     1     2     3     4         5     6     7     8     9    10    11    12        13    14        15    16        17    18    19    20    21    22        23    24    25    26    27    28        29    30        31    32    33    34    35        36        37        38        39        40    41        42    43    44    45    46        47    48    49        50        51    52        53        54    55    56    57        58        59        60        61    62        63        64  \
tweet                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
0       0.28  0.65 -0.16 -0.33  0.02 -1.45e-02  0.48 -0.15  0.24  0.06 -0.12 -0.09 -0.14  1.22e-01  0.19  3.11e-03  0.07 -2.62e-02  0.25  0.69  0.42  0.10  0.35 -5.62e-02  0.10 -0.03 -0.02  0.01 -0.08 -1.29e-01 -0.14  1.70e-04  0.09  0.02 -0.11  0.13  9.80e-03  4.33e-03  1.14e-01  2.40e-02  8.73e-03 -0.06 -3.55e-02  0.06  0.06 -0.03  0.04  7.47e-04 -0.02 -0.06 -1.03e-01 -1.84e-02 -0.06 -9.53e-03 -4.94e-02  0.07  0.09 -0.06 -6.50e-02 -1.85e-02  9.49e-02 -8.60e-03  0.02  6.24e-02  1.34e-02   
1       0.66 -0.58  0.15  0.06 -0.04  1.74e-04  0.17 -0.13  0.06 -0.02 -0.25 -0.07 -0.17  1.06e-02  0.06 -8.07e-02 -0.08  6.70e-03 -0.09  0.03  0.07  0.03  0.05 -2.42e-02 -0.08 -0.01  0.03 -0.02  0.04  2.87e-05  0.04  5.37e-02 -0.01 -0.01  0.03  0.05 -1.70e-02 -1.52e-02  8.76e-03 -1.44e-03 -1.54e-02 -0.04  5.26e-02  0.08  0.03 -0.02  0.04 -1.76e-02 -0.03 -0.02 -2.00e-02 -8.98e-03 -0.05  2.13e-02  1.85e-02  0.08  0.02 -0.03 -1.43e-03  5.79e-02  1.35e-03 -5.04e-02 -0.08  2.39e-03  5.05e-02   
2       0.73 -0.59  0.15  0.08 -0.06 -1.07e-02  0.20 -0.16  0.06 -0.02 -0.34 -0.07 -0.24  5.64e-03  0.11 -3.67e-02 -0.09 -6.77e-02 -0.17 -0.05  0.07  0.05  0.04 -2.44e-01 -0.09 -0.03 -0.03 -0.01 -0.18 -3.83e-01 -0.01 -2.29e-01  0.10  0.34  0.17 -0.46  5.17e-02 -4.79e-03  1.52e-01 -2.18e-02 -1.10e-01 -0.03  2.36e-02  0.37 -0.14  0.19  0.20 -5.82e-01  0.58  0.34 -3.31e-02  1.08e-01 -0.12 -5.86e-02  1.21e-01 -0.08  0.04 -0.08 -2.00e-01  5.30e-03 -1.97e-01 -4.81e-02  0.26  4.30e-02  5.47e-03   
...      ...   ...   ...   ...   ...       ...   ...   ...   ...   ...   ...   ...   ...       ...   ...       ...   ...       ...   ...   ...   ...   ...   ...       ...   ...   ...   ...   ...   ...       ...   ...       ...   ...   ...   ...   ...       ...       ...       ...       ...       ...   ...       ...   ...   ...   ...   ...       ...   ...   ...       ...       ...   ...       ...       ...   ...   ...   ...       ...       ...       ...       ...   ...       ...       ...   
183067  1.43  0.12 -0.84 -0.15 -0.25 -1.56e-01  0.27 -0.58  0.48 -0.18 -1.24  0.63  1.15  4.95e-01 -0.62  1.38e-01  0.13  1.91e-01 -0.05  0.05 -0.02  0.22  0.21 -3.18e-02 -0.08 -0.18  0.09  0.04  0.06 -6.57e-02  0.08  1.47e-01  0.02  0.06  0.04  0.10 -4.27e-02 -2.52e-02  6.43e-02  9.56e-02 -8.67e-02  0.06  3.82e-03  0.16  0.05  0.02  0.09  3.79e-02 -0.09 -0.05  1.00e-03  1.14e-02  0.03  1.36e-02 -8.07e-02  0.04  0.03 -0.07  9.18e-03 -5.47e-02 -2.22e-02 -1.41e-02  0.01 -1.69e-02  4.97e-02   
183068  0.66 -0.58  0.15  0.06 -0.04  1.74e-04  0.17 -0.13  0.06 -0.02 -0.25 -0.07 -0.17  1.06e-02  0.06 -8.07e-02 -0.08  6.70e-03 -0.09  0.03  0.07  0.03  0.05 -2.42e-02 -0.08 -0.01  0.03 -0.02  0.04  2.87e-05  0.04  5.37e-02 -0.01 -0.01  0.03  0.05 -1.70e-02 -1.52e-02  8.76e-03 -1.44e-03 -1.54e-02 -0.04  5.26e-02  0.08  0.03 -0.02  0.04 -1.76e-02 -0.03 -0.02 -2.00e-02 -8.98e-03 -0.05  2.13e-02  1.85e-02  0.08  0.02 -0.03 -1.43e-03  5.79e-02  1.35e-03 -5.04e-02 -0.08  2.39e-03  5.05e-02   
183069  0.68 -0.61  0.17 -0.02 -0.03 -4.62e

In [154]:
dates = pd.read_csv(os.path.join(DATA_PATH, 'datetimes.csv.gz'), engine='python')
nums = pd.read_csv(os.path.join(DATA_PATH, 'numbers.csv.gz'), engine='python')

When I first ran this, my dataframes weren't "aligned".  
So it's very important to check your datasets after every load.  
The correspondence between dates and topics and numerical features is critical for training!  

In [155]:
print(len(dates))
print(len(topics))
print(len(nums))

183070
183070
183070


In [156]:
sum(nums.index == dates.index) == len(dates)

True

In [157]:
sum(nums.index == topics.index) == len(dates)

True

In [158]:
disc = LinearDiscriminantAnalysis()
disc

LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [159]:
features = pd.concat((topics, nums.favorite_count), axis=1)
features

0     1     2     3     4         5     6     7     8     9    10    11    12        13    14        15    16        17    18    19    20    21    22        23    24    25    26    27    28        29    30        31    32    33    34    35        36        37        38        39        40    41        42    43    44    45    46        47    48    49        50        51    52        53        54    55    56    57        58        59        60        61    62        63        64  \
tweet                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
0       0.28  0.65 -0.16 -0.33  0.02 -1.45e-02  0.48 -0.15  0.24  0.06 -0.12 -0.09 -0.14  1.22e-01  0.19  3.11e-03  0.07 -2.62e-02  0.25  0.69  0.42  0.10  0.35 -5.62e-02  0.10 -0.03 -0.02  0.01 -0.08 -1.29e-01 -0.14  1.70e-04  0.09  0.02 -0.11  0.13  9.80e-03  4.33e-03  1.14e-01  2.40e-02  8.73e-03 -0.06 -3.55e-02  0.06  0.06 -0.03  0.04  7.47e-04 -0.02 -0.06 -1.03e-01 -1.84e-02 -0.06 -9.53e-03 -4.94e-02  0.07  0.09 -0.06 -6.50e-02 -1.85e-02  9.49e-02 -8.60e-03  0.02  6.24e-02  1.34e-02   
1       0.66 -0.58  0.15  0.06 -0.04  1.74e-04  0.17 -0.13  0.06 -0.02 -0.25 -0.07 -0.17  1.06e-02  0.06 -8.07e-02 -0.08  6.70e-03 -0.09  0.03  0.07  0.03  0.05 -2.42e-02 -0.08 -0.01  0.03 -0.02  0.04  2.87e-05  0.04  5.37e-02 -0.01 -0.01  0.03  0.05 -1.70e-02 -1.52e-02  8.76e-03 -1.44e-03 -1.54e-02 -0.04  5.26e-02  0.08  0.03 -0.02  0.04 -1.76e-02 -0.03 -0.02 -2.00e-02 -8.98e-03 -0.05  2.13e-02  1.85e-02  0.08  0.02 -0.03 -1.43e-03  5.79e-02  1.35e-03 -5.04e-02 -0.08  2.39e-03  5.05e-02   
2       0.73 -0.59  0.15  0.08 -0.06 -1.07e-02  0.20 -0.16  0.06 -0.02 -0.34 -0.07 -0.24  5.64e-03  0.11 -3.67e-02 -0.09 -6.77e-02 -0.17 -0.05  0.07  0.05  0.04 -2.44e-01 -0.09 -0.03 -0.03 -0.01 -0.18 -3.83e-01 -0.01 -2.29e-01  0.10  0.34  0.17 -0.46  5.17e-02 -4.79e-03  1.52e-01 -2.18e-02 -1.10e-01 -0.03  2.36e-02  0.37 -0.14  0.19  0.20 -5.82e-01  0.58  0.34 -3.31e-02  1.08e-01 -0.12 -5.86e-02  1.21e-01 -0.08  0.04 -0.08 -2.00e-01  5.30e-03 -1.97e-01 -4.81e-02  0.26  4.30e-02  5.47e-03   
...      ...   ...   ...   ...   ...       ...   ...   ...   ...   ...   ...   ...   ...       ...   ...       ...   ...       ...   ...   ...   ...   ...   ...       ...   ...   ...   ...   ...   ...       ...   ...       ...   ...   ...   ...   ...       ...       ...       ...       ...       ...   ...       ...   ...   ...   ...   ...       ...   ...   ...       ...       ...   ...       ...       ...   ...   ...   ...       ...       ...       ...       ...   ...       ...       ...   
183067  1.43  0.12 -0.84 -0.15 -0.25 -1.56e-01  0.27 -0.58  0.48 -0.18 -1.24  0.63  1.15  4.95e-01 -0.62  1.38e-01  0.13  1.91e-01 -0.05  0.05 -0.02  0.22  0.21 -3.18e-02 -0.08 -0.18  0.09  0.04  0.06 -6.57e-02  0.08  1.47e-01  0.02  0.06  0.04  0.10 -4.27e-02 -2.52e-02  6.43e-02  9.56e-02 -8.67e-02  0.06  3.82e-03  0.16  0.05  0.02  0.09  3.79e-02 -0.09 -0.05  1.00e-03  1.14e-02  0.03  1.36e-02 -8.07e-02  0.04  0.03 -0.07  9.18e-03 -5.47e-02 -2.22e-02 -1.41e-02  0.01 -1.69e-02  4.97e-02   
183068  0.66 -0.58  0.15  0.06 -0.04  1.74e-04  0.17 -0.13  0.06 -0.02 -0.25 -0.07 -0.17  1.06e-02  0.06 -8.07e-02 -0.08  6.70e-03 -0.09  0.03  0.07  0.03  0.05 -2.42e-02 -0.08 -0.01  0.03 -0.02  0.04  2.87e-05  0.04  5.37e-02 -0.01 -0.01  0.03  0.05 -1.70e-02 -1.52e-02  8.76e-03 -1.44e-03 -1.54e-02 -0.04  5.26e-02  0.08  0.03 -0.02  0.04 -1.76e-02 -0.03 -0.02 -2.00e-02 -8.98e-03 -0.05  2.13e-02  1.85e-02  0.08  0.02 -0.03 -1.43e-03  5.79e-02  1.35e-03 -5.04e-02 -0.08  2.39e-03  5.05e-02   
183069  0.68 -0.61  0.17 -0.02 -0.03 -4.62e

In [160]:
disc = LinearDiscriminantAnalysis().fit(topics, nums.favorite_count >= 1)

In [161]:
predicted_favorites = disc.predict(topics)
predicted_favorites

array([False, False, False, ..., False, False, False], dtype=bool)

In [162]:
np.sum(predicted_favorites)

4228

## Wow!  
DiscriminantAnalysis is VERY discriminating!

In [163]:
np.sum(nums.favorite_count >= 1)

33292

But not in a good way.  
10x more true favorites than predicted. 
Our unbalanced training set makes it easy for the judge to be tough.  
Let's mellow our judge a bit...  

In [164]:
from pug.nlp.stats import Confusion

In [165]:
results = pd.DataFrame()
results['predicted'] = predicted_favorites
results['truth'] = pd.Series(nums.favorite_count >= 1)
conf = Confusion(results)
conf

truth       False  True 
predicted               
False      147725  31117
True         2053   2175

In [166]:
results.predicted.corr(results.truth)

0.1326

In [167]:
conf.stats_dict

PrettyDict([('tpr', 0.8260),
            ('fpr', 0.4856),
            ('tnr', 0.5144),
            ('fnr', 0.1740),
            ('plr', 1.7011),
            ('nlr', 0.3382),
            ('accuracy', 0.8188),
            ('sensitivity', 0.8260),
            ('specificity', 0.5144),
            ('mcc', 0.1326),
            ('chi_squared', 3217.3986)])

High accuracy, but low MCC (correlation)

Balance the training?  
Get rid of some negatives?  
**Accentuate the positive?** <-- give this a try yourself

In [168]:
pos = np.array(nums.favorite_count >= 1)
neg = ~pos
portion_pos = float(sum(pos)) / len(nums)
mask = ((np.random.binomial(1, portion_pos, size=len(nums)).astype(bool) & neg) | pos)
disc = LinearDiscriminantAnalysis().fit(topics[mask], (nums.favorite_count[mask] >= 1))
print(sum(mask))
print(sum(pos) * 2)

60633
66584


In [169]:
results = pd.DataFrame()
results['predicted'] = disc.predict(topics.values)
results['truth'] = nums.favorite_count.values >= 1
conf = Confusion(results)
conf

truth      False  True 
predicted              
False      79751   1355
True       70027  31937

In [170]:
results.predicted.corr(results.truth)

0.3819

In [171]:
conf.stats_dict

PrettyDict([('tpr', 0.9833),
            ('fpr', 0.6868),
            ('tnr', 0.3132),
            ('fnr', 0.0167),
            ('plr', 1.4317),
            ('nlr', 0.0533),
            ('accuracy', 0.6101),
            ('sensitivity', 0.9833),
            ('specificity', 0.3132),
            ('mcc', 1.1766),
            ('chi_squared', 253427.2867)])

So let's add some more negative examples back in.  
50x imbalance is defintiely misleading.  
But 2-5x imbalance is probably OK.  

In [172]:
portion_neg = 3 * portion_pos
mask = ((np.random.binomial(1, portion_neg, size=len(nums)).astype(bool) & neg) | pos)
disc = LinearDiscriminantAnalysis().fit(topics[mask], nums.favorite_count[mask] >=1 )
print(sum(mask))
print(sum(pos) * 2)

114711
66584


In [173]:
results = pd.DataFrame()
results['predicted'] = disc.predict(topics.values)
results['truth'] = nums.favorite_count.values > 0
conf = Confusion(results)
conf

truth       False  True 
predicted               
False      133025  20356
True        16753  12936

At least the confusion matrix looks balanced now

In [174]:
results.predicted.corr(results.truth)

0.2896

Should have known, imbalance doesn't help...

In [179]:
portion_neg = 2 * portion_pos
mask = ((np.random.binomial(1, portion_neg, size=len(nums)).astype(bool) & neg) | pos)
disc = LinearDiscriminantAnalysis().fit(topics.values[mask], (nums.favorite_count.values > 0)[mask])
print(sum(mask))
print(sum(pos) * 2)

87655
66584


In [180]:
results = pd.DataFrame()
results['predicted'] = disc.predict(topics.values)
results['truth'] = nums.favorite_count.values > 0
conf = Confusion(results)
conf

truth       False  True 
predicted               
False      101101   6434
True        48677  26858

In [181]:
results.predicted.corr(results.truth)

0.3775

So it looks like 38% correlation is all we can squeeze out of this simple model  
Next up... adding number of followers and other features  